src/DataFramesMeta.jl

```diff
@@ -388,19 +387,86 @@ end
 
 function transform(g::GroupedDataFrame; kwargs...)
     result = DataFrame(g)
-    idx2 = cumsum(Int[size(g[i],1) for i in 1:length(g)])
-    idx1 = [1; 1 .+ idx2[1:end-1]]
+    ends = cumsum(Int[size(g[i],1) for i in 1:length(g)])
+    starts = [1; 1 .+ ends[1:end-1]]
+    lengths = [ends[i] - starts[i] + 1 for i in 1:length(starts)]
     for (k, v) in kwargs
         first = v(g[1])
-        result[k] = Array{eltype(first)}(undef, size(result, 1))
-        result[idx1[1]:idx2[1], k] = first
-        for i in 2:length(g)
-            result[idx1[i]:idx2[i], k] = v(g[i])
+        if first isa AbstractVector
+            t = _transform!(Tables.allocatecolumn(eltype(first), size(result, 1)), first, 1, g, v, starts, ends)
+        else 
+            t = _transform!(Tables.allocatecolumn(typeof(first), size(result, 1)), first, 1, g, v, starts, ends)
         end
+        result[k] = t
     end
     return result
 end
 
+function _transform!(t::AbstractVector, first::AbstractVector, start::Int, g::GroupedDataFrame, v::Function, starts::Vector, ends::Vector)
+    # handle the first case 
+    j = fill_column_vec!(t, first, starts[start], ends[start], size(g[start], 1))
```

Comment:
`j` is a bit weird here. I used that name in the other PR because it was an index. Maybe `promoted` just like in the other place? Or even better `newT`/`newtype` (in both places): `promoted` sounds like a Boolean.